packages/core/src/sync/index.ts

```diff
@@ -936,243 +817,442 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
+          common.metrics.ponder_sync_is_realtime.set(
+            { network: network.name },
+            0,
+          );
+          common.metrics.ponder_sync_is_complete.set(
+            { network: network.name },
+            1,
+          );
+          common.logger.info({
+            service: "sync",
+            msg: `Killing '${network.name}' live indexing because the end block ${hexToNumber(syncProgress.end!.number)} has been finalized`,
+          });
+          realtimeSync.kill();
+        }
+
+        return event;
+      }
+
+      case "reorg": {
+        syncProgress.current = event.block;
+
+        common.logger.debug({
+          service: "sync",
+          msg: `Updated '${network.name}' current block to ${hexToNumber(event.block.number)}`,
+        });
+
+        common.metrics.ponder_sync_block.set(
+          { network: network.name },
+          hexToNumber(syncProgress.current.number),
+        );
+
+        // Remove all reorged data
+
+        unfinalizedBlocks = unfinalizedBlocks.filter(
+          ({ block }) =>
+            hexToNumber(block.number) <= hexToNumber(event.block.number),
+        );
+
+        await syncStore.pruneRpcRequestResult({
+          chainId: network.chainId,
+          blocks: event.reorgedBlocks,
+        });
+
+        return event;
+      }
+    }
   };
 };
 
-/** Returns the closest-to-tip block that has been synced for all `sources`. */
-export const getCachedBlock = ({
-  sources,
-  requestQueue,
-  historicalSync,
-}: {
+export async function* getLocalEventGenerator(params: {
+  common: Common;
+  network: Network;
+  syncStore: SyncStore;
   sources: Source[];
-  requestQueue: RequestQueue;
-  historicalSync: HistoricalSync;
-}): Promise<SyncBlock | LightBlock> | undefined => {
-  const latestCompletedBlocks = sources.map(({ filter }) => {
-    const requiredInterval = [
-      filter.fromBlock ?? 0,
-      filter.toBlock ?? Number.POSITIVE_INFINITY,
-    ] satisfies Interval;
-    const fragmentIntervals = historicalSync.intervalsCache.get(filter)!;
+  localSyncGenerator: AsyncGenerator<string>;
+  from: string;
+  to: string;
+  limit: number;
+}): AsyncGenerator<{ events: RawEvent[]; checkpoint: string }> {
+  let cursor = params.from;
+  // Estimate optimal range (seconds) to query at a time, eventually
+  // used to determine `to` passed to `getEvents`.
+  let estimateSeconds = 1_000;
 
-    const completedIntervals = sortIntervals(
-      intervalIntersection(
-        [requiredInterval],
-        intervalIntersectionMany(
-          fragmentIntervals.map(({ intervals }) => intervals),
+  params.common.logger.debug({
+    service: "sync",
+    msg: `Initialized '${params.network.name}' extract query for timestamps [${decodeCheckpoint(params.from).blockTimestamp}, ${decodeCheckpoint(params.to).blockTimestamp}]`,
+  });
+
+  for await (const syncCheckpoint of bufferAsyncGenerator(
+    params.localSyncGenerator,
+    Number.POSITIVE_INFINITY,
```

Comment:
So, this is what tells the historical sync to just keep cooking regardless of indexing progress?

Reply:
yes